dbms/src/Storages/KVStore/MultiRaft/ApplySnapshot.cpp

```diff
@@ -80,8 +80,8 @@ void KVStore::checkAndApplyPreHandledSnapshot(const RegionPtrWrap & new_region,
             // engine may delete data unsafely.
             auto region_lock = region_manager.genRegionTaskLock(old_region->id());
             old_region->setStateApplying();
-            tmt.getRegionTable().tryWriteBlockByRegion(old_region);
-            tryFlushRegionCacheInStorage(tmt, *old_region, log);
+            // It is not worthy to call `tryWriteBlockByRegion` and `tryFlushRegionCacheInStorage` here,
+            // even if the written data is useful, it could be overwritten later in `onSnapshot`.
             persistRegion(*old_region, region_lock, PersistRegionReason::ApplySnapshotPrevRegion, "");
```

Comment:
If crash happen after line 85 but before the apply snapshot is fully done. It is safe?
I remember that if a region is under state "applying", then the read index to the region will be treat as not exist?

Reply:
If `onSnapshot` is called, then the snapshot is already applied in the aspect of raftstore.
The applied snapshot could not be revoked, if there is any failure in `onSnapshot`. So if the crash happens after line 85 and before the snapshot is finished, TiFlash will retry the snapshot when it is restarted. It could crash over and over again, until manually interfere.
An `Applying` region can not be read.